### ISSUE TITLE
Relative imports from conanfile.py don't work

### ISSUE THREAD

**Issue by owbone** [conan-io/conan#39]
I have a Python module to separate out some code which would otherwise be duplicated across a bunch of `conanfile.py`s. It lives next to each conanfile like so:
```
+ .
|--+ conan_helpers
|  |-- __init__.py
|  \-- ...
\--- conanfile.py
```

And the `conanfile.py` imports it as normal:
```python
from conans import ConanFile
from conan_helpers import SomeMixin

class SomeConan(SomeMixin, ConanFile):
    pass
```

However this fails during `conan export` with an `ImportError`:
```
ImportError: No module named conan_helpers
```

The workaround at the moment is to add the directory of `__file__` to `sys.path` before the import statement, but it would be nice if I didn't have to do that. ;)

**Comment by memsharded** [comment]
Hi @owbone!

I have been playing a little bit with this concept, but I must admit that I can't see a good solution. Check: https://github.com/conan-io/conan/compare/develop...memsharded:feature/python_helpers?expand=1

You can see ConanfileHelpersTest, this test actually passes, even it is not "exporting" the helpers files. The reason seems to be the module importing mechanism in python, so the second comand "build" is able to locate the files in the original directory! This is wrong, and shouldn't be allowed. Even if in an actual run, it will run in 2 commands, with separate sys.modules instances, there is a high risk of problems between different packages importing the same "helper" module, that it will not be the local, relative one, but one previously imported by other conanfile.py.

If you have any idea how this could be fixed, very welcomed :) Otherwise, I think it could be better not to merge this behavior, and fallback to old copy&paste. Conanfiles.py would have some duplication, but will always be self-contained and totally explicit which is a good thing, and we won't be debugging our conanfiles for some strange cross interaction with another conanfile importing the same helper code. What do you think?

**Comment by owbone** [comment]
Hi @memsharded,

Thanks for the quick investigation.

> You can see ConanfileHelpersTest, this test actually passes, even it is not "exporting" the helpers files. The reason seems to be the module importing mechanism in python, so the second comand "build" is able to locate the files in the original directory! This is wrong, and shouldn't be allowed.

I think we should work out how this is happening first, since it sounds like you've just exposed a separate bug if the build step isn't failing. There are maybe three possibilities I can think of right now:

 * conan is using the local conanfile.py before the exported conanfile.py when building.
 * The helpers module is somehow being exported along with the conanfile.py.
 * Python is searching for imports in the local directory.
 * Something else?

The first one seems the most likely, and I assume that would be a separate bug in conan. It could be proved or disproved easily by changing the directory of `client2` in your test code.

> Even if in an actual run, it will run in 2 commands, with separate sys.modules instances, there is a high risk of problems between different packages importing the same "helper" module, that it will not be the local, relative one, but one previously imported by other conanfile.py.

Is that something which conan should be concerned with though? You could say the same thing about any two python scripts. I think the above problem is confusing things here.

> If you have any idea how this could be fixed, very welcomed :) Otherwise, I think it could be better not to merge this behavior, and fallback to old copy&paste. Conanfiles.py would have some duplication, but will always be self-contained and totally explicit which is a good thing, and we won't be debugging our conanfiles for some strange cross interaction with another conanfile importing the same helper code. What do you think?

As mentioned above, any cross interaction is a bug in conan which should be fixed and would otherwise not be a problem.

I don't think limiting people's code into a single python script is a good idea in the long run. One of the nice things about conan is the ability to wrap arbitrarily complicated source control and build systems, and sometimes they *are* complicated enough that the solution should be made properly reusable.

For example, I think it would make a company more willing to use conan if it could wrap some stupid, proprietary source control system and simplify it down to something like this...
```python
    def source(self):
        some_source_control.checkout('some_module')
```
...and then reuse that solution for all of their conan packages. Not only does it make the code nicer but conan would be allowing them to write proper abstractions which could enable them to switch source control systems without rewriting all of their packages. These kinds of things are important to businesses who want to waste as little time as possible.

My current preferred solution would be:

 1. Fix the bug which is allowing the `install` command to find the module in the wrong directory.
 2. Have the `export` command recursively export the whole directory rather than just the conanfile.

I don't really know enough about conan to know how possible exporting the entire directory would be.

**Comment by memsharded** [comment]
Regarding 2, it is quite simple, it can be easily done with the exports class attribute:

```python
class SomeLib(Conanfile):
    exports = "*"
```

However, this is not the desired behavior in the general case, as the main use case is to retrieve the source code from some VCS or download. Furthermore, exporting all the files doesn't solve the issue, the actual code used is the one imported by python modules system at first, not the one exported by conan together with the conanfile.py.

So, yes, we should further investigate 1. Let me make sure that I understood it right, the python code implementing for example ```some_source_control.checkout```, would be exported and uploaded together with the conanfile.py, right? So developers of the respective package would only need to copy some python files into the package conanfile.py folder. 

I will have another look to the python module loading process in a couple of days, after the 0.5 release, to see how this problem can be solved. Thanks for your detailed analysis!

**Comment by owbone** [comment]
> Regarding 2, it is quite simple, it can be easily done with the exports class attribute:

Thanks. I'll try that!

> Furthermore, exporting all the files doesn't solve the issue, the actual code used is the one imported by python modules system at first, not the one exported by conan together with the conanfile.py.

Python will *always* import from the current script's directory first, so as long as everything is exported together then it will *always* import the correct one. I'm pretty sure that the only reason you were seeing differently is because of the other bug.

> So, yes, we should further investigate 1. Let me make sure that I understood it right, the python code implementing for example some_source_control.checkout, would be exported and uploaded together with the conanfile.py, right? So developers of the respective package would only need to copy some python files into the package conanfile.py folder.

That's right, and it would automatically allow using something nicer like git submodules rather than copying the files by hand.

**Comment by memsharded** [comment]
I have updated the branch, you can check the approach here: https://github.com/conan-io/conan/compare/develop...memsharded:feature/python_helpers?expand=1

This seems to work, it is able to import the different helpers, even if the modules are equally named
The helpers have to be "exported" by conan, so the conanfile.py recipe can import them when necessary, irrespective of the folder (source, build, package)

The implementation hacks the sys.modules dict. The problem is that each conanfile is not inside a named python package, because in fact they are not part of a python project, no __init__.py files... just s conanfile.py that has to be loaded. So we were defining a new module name based on UUID for each loaded conanfile. Now, we compare the sys.modules before and after loading each conanfile. All the new imported modules that are in the conanfile.py directory, are considered to be part of that conanfile.py package, and both the conanfile and all those local modules have their names in sys.modules prefixed with the UUID. 

I am not particularly comfortable with this solution, but it passes the tests, and has no apparent problems, but for example, in the comparison of sys.modules, it is necessary to define a set(sys.modules) for the difference, otherwise it fails. It seems that the order in which the new modules are processed, affects the consistency of further processing. It seems that set(sys.modules) gives a good ordering, but I don't fully understand why yet.

What do you think about this approach? Any other better idea?

**Comment by owbone** [comment]
I don't really understand what problem you're trying to solve now.

> The problem is that each conanfile is not inside a named python package, because in fact they are not part of a python project, no init.py files... just s conanfile.py that has to be loaded.

Why is that a problem?

> So we were defining a new module name based on UUID for each loaded conanfile. 

Were you? From what I can see, the module name is hardcoded to `"conanfile"` and `module_id` is an unused variable.

> Now, we compare the sys.modules before and after loading each conanfile. All the new imported modules that are in the conanfile.py directory, are considered to be part of that conanfile.py package, and both the conanfile and all those local modules have their names in sys.modules prefixed with the UUID.

Why? What does that achieve?

A combination of your previous change (just the proper directory name adding to `sys.path`) and `exports='*'` seems *much* more correct, and you've lost me as to why that isn't sufficient.

FYI, a Python `set` is ordered by definition.

**Comment by memsharded** [comment]
There is a bunch of conanfile.py, all with the same name, living in different folders in the filesystem, the conan storage (where the binary packages live), the user projects, etc. They are not inside a python project, a python package, they dont have _ _init_ _.py. So if we import them as:
```python
loaded = imp.load_source("conanfile", conan_file_path)
```
The "conanfile"  name will be added to ```sys.modules``` and avoid loading multiple different conanfiles. In the past version, an uuid (a new one per load) to guarantee uniqueness was added to the name:

```python
loaded = imp.load_source("conanfile%s" % uuid, conan_file_path)
```
In that way we avoided name collisions inside ```sys.modules``` for all conanfiles, that were named ```conanfileUUID1```, ```conanfileUUID2``` (replace UUIDX with actual UUIDs)

Now, the new proposed solution creates a package-like name for all the modules loaded in the last ```imp.load_source```. The ```uuid``` named ```module_id``` is used, check it:

```python
for added in added_modules:
    m = sys.modules[added]
    folder = os.path.dirname(m.__file__)
    if folder.startswith(current_dir):
        m = sys.modules.pop(added)
        sys.modules["%s.conanfile" % module_id] = m  # HERE
```

With this approach, the conanfiles modules are named ```UUID1.conanfile```, ```UUID2.conanfile```, as if they lived in different packages. This approach can be extended to other modules loaded by the ```imp.load_source```, all modules that live in the same folder are also prefixed with the same UUID, as the conanfile as ``` UUID1.helper```, ```UUID1.helpers.other```, while maintaining other system modules (os, sys...) not prefixed. This allow other conanfiles, from other libraries, to have their own local imports even with the same names, which would be named ``` UUID2.helper```, ```UUID2.helpers.other``` and will not collide with the other ones.

Adding to ```sys.path``` and exporting with ```exports="*"``` is the proposed solution, it is the correct one, but it is not sufficient: Note the ConanfileHelpersTest now has been augmented with external helpers for 2 packages, and are imported from their respective conanfile with the same import, but they must have different module names because they are indeed different modules with different contents, so it is necessary that they have a different name in ```sys.modules```, and this is why I have prefixed them with the same prefix as the conanfile that they support.

From the python official docs: https://docs.python.org/2/library/stdtypes.html#set

    A set object is an unordered collection of distinct hashable objects. 

So, no, a python set is not ordered by definition. 

That is why I have concerns about some weird failures in the test when ```old_modules = sys.modules.keys()``` is used instead of ```set(sys.modules)```. It shouldn't be a big problem, as it could potentially affect only to the users that use this import feature, but the remaining package and users should be fine.

**Comment by owbone** [comment]
Thanks for the explanation, I think the part of the problem I wasn't aware of was loading *multiple conan files at the same time*. Is that something that conan still does? Surely not since the UUID wasn't being used any more and they're all imported as "conanfile"?

> Adding to sys.path and exporting with exports="*" is the proposed solution, it is the correct one, but it is not sufficient: Note the ConanfileHelpersTest now has been augmented with external helpers for 2 packages, and are imported from their respective conanfile with the same import, but they must have different module names because they are indeed different modules with different contents, so it is necessary that they have a different name in sys.modules, and this is why I have prefixed them with the same prefix as the conanfile that they support.

So let's see if I understand at this point:

 * The fix **is** sufficient for conan itself, since conan will not import more than one conanfile at a time.
 * The fix **is not** sufficient for ConanfileHelpersTest because that *does* currently import more than one conanfile.

If both are true then there's something fundamentally wrong here because ConanfileHelpersTest is creating a problem which doesn't exist in the application itself. If that's the case then the original fix *is* sufficient but the test is doing the wrong thing.

> So, no, a python set is not ordered by definition.

Sorry, that was me confusing it with a C++ `set` based on the ordering you were seeing. I definitely wouldn't rely on that ordering then if I were you, since it's undefined so could easily be different between different OSs, computers, Python versions, etc, so it would be another bug waiting to happen.

**Comment by memsharded** [comment]
You are welcome. :) Yes, that was the issue, conan is a dependency manager, that handles dependency override, dependency conflict resolution, conditional dependencies... So it has to load the full directed graph of all dependencies, which is a graph that has a conanfile at each node, and each node has directional edges to its dependencies. Each conanfile comes from a different package (even two conanfiles of the same package, different versions can be loaded some times when handling with conflicts, so the name of the class is not valid either to differentiate), and they share information once loaded (options, requirements, build_information), so processing them one by one is not an option.

So far, the fix might be sufficient for conan itself, and it is sufficient for the ConanfileHelpersTest, it can manage importing auxiliary python files for different conanfiles at the same time, and the test passes.

About the set, yes, that kind of stuff, mixing features from different languages, happens a lot when you switch between them often, I know it by myself, no need to say sorry :)

So the solution seems to be properly outlined and might be ok, but as you said, I won't rely on it until fully understanding why iterating over the ```sys.modules.keys()``` of the modules fails, and using a ```set(sys.modules)``` works ok, particularly because for this problem the order shouldn't matter. It is on my todo for 0.6, lets hope we can release 0.5 soon, and allocate more time for this. Thanks again!

**Comment by memsharded** [comment]
Yes, I got it! :) :) :) 

It was a bug, one of my stupid mistakes, instead of recreating all relatively imported modules names with their respective prefixed names I was doing:

```python
sys.modules["%s.conanfile" % module_id] = m
```
So the ordering was really important, because all the modules were being renamed to UUIDX.conanfile. Just by (bad luck) chance the test was passing (if the conanfile is the last module to be renamed, it works). I generated random names in the test for the imports, and the bug appeared. Easy to fix with:

```python
sys.modules["%s.%s" % (module_id, added)] = m
```

Now everything works like a charm, tested with large number of random import names, and everything ok, so I submitted a PR: https://github.com/conan-io/conan/pull/53

It should be in next release 0.6. Thanks very much for your help, I hope this is a useful feature.
